src/plugins/rpd/rpdplugin.cpp

```diff
@@ -0,0 +1,319 @@
+#include <QJsonDocument>
+#include <QJsonObject>
+#include <QTextStream>
+#include <QtMath>
+
+namespace Rpd {
+
+void RpdPlugin::initialize()
+{
+    addObject(new RpdMapFormat(RpdMapFormat::Rpd, this));
+    addObject(new RpdTilesetFormat(this));
+}
+
+RpdMapFormat::RpdMapFormat(SubFormat subFormat, QObject *parent)
+    : Tiled::WritableMapFormat(parent)
+    , mSubFormat(subFormat)
+{}
+
+QString RpdMapFormat::shortName() const
+{
+    return "RPD";
+}
+
+static QJsonArray packMapData(Tiled::TileLayer *layer)
+{
+    QJsonArray map;
+    for (int j = 0; j < layer->map()->height(); ++j)
+        for (int i = 0; i < layer->map()->width(); ++i)
+            map.append(layer->cellAt(i, j).tileId());
+    return map;
+}
+
+bool RpdMapFormat::insertTilesetFile(Tiled::Layer &layer,
+                                     const QString &tiles_name,
+                                     QJsonObject &mapJson)
+{
+    const auto tilesets = layer.usedTilesets();
+
+    if (tilesets.isEmpty()) {
+        mError = tr("You have an empty %1 layer, please fill it").arg(layer.name());
+        return false;
+    }
+
+    if (tilesets.size() > 1) {
+        mError = tr("Only one tileset per layer supported (%1 layer) ->\n").arg(layer.name());
+        for (const auto &tileset : tilesets)
+            mError += "[" + tileset->name() + "]\n";
+        return false;
+    }
+
+    mapJson.insert(tiles_name, tilesets.begin()->data()->name() + ".png");
+    return true;
+}
+
+bool RpdMapFormat::write(const Tiled::Map *map, const QString &fileName, Options options)
+{
+    Tiled::SaveFile file(fileName);
+
+    if (!file.open(QIODevice::WriteOnly | QIODevice::Text)) {
+        mError = tr("Could not open file for writing.");
+        return false;
+    }
+
+    QJsonObject mapJson;
+
+    for (const auto &[key, value] : map->properties().toStdMap()) {
+        if (value.canConvert<double>()) {
+            mapJson.insert(key, value.toDouble());
+            continue;
+        }
+
+        if (value.canConvert<QString>()) {
+            mapJson.insert(key, value.toString());
+            continue;
+        }
+
+        mError = tr("Dont know what to do with property (%1)").arg(key);
+        return false;
+    }
+
+    for (Tiled::Layer *layer : map->layers()) {
+        if (layer->layerType() == Tiled::Layer::TileLayerType
+            && !insertTilesetFile(*layer, QString("tiles_") + layer->name(), mapJson)) {
+            return false;
+        }
+
+        if (layer->name() == "logic") {
+            QJsonArray entrance;
+            QJsonArray multiexit;
+
+            mapJson.insert("width", layer->map()->width());
+            mapJson.insert("height", layer->map()->height());
+
+            mapJson.insert("map", packMapData(layer->asTileLayer()));
+
+            for (int i = 0; i < layer->map()->width(); ++i) {
+                for (int j = 0; j < layer->map()->height(); ++j) {
+                    int tileId = layer->asTileLayer()->cellAt(i, j).tileId();
```

Comment:
Plugin will crash if exporting a map with a layer named "logic" that isn't a tile layer. Layer type needs to be checked along with the name above (same problem with other layers below).